Source/OxyPlot.Wpf/CanvasRenderContext.cs

```diff
@@ -136,20 +107,43 @@ public override void DrawEllipse(OxyRect rect, OxyColor fill, OxyColor stroke, d
         ///<inheritdoc/>
         public override void DrawEllipses(IList<OxyRect> rectangles, OxyColor fill, OxyColor stroke, double thickness, EdgeRenderingMode edgeRenderingMode)
         {
+            if (rectangles.Count == 0)
+            {
+                return;
+            }
+
             var path = this.CreateAndAdd<Path>();
             this.SetStroke(path, stroke, thickness, edgeRenderingMode);
             if (!fill.IsUndefined())
             {
                 path.Fill = this.GetCachedBrush(fill);
             }
 
-            var gg = new GeometryGroup { FillRule = FillRule.Nonzero };
-            foreach (var rect in rectangles)
+            var isFilled = !fill.IsUndefined();
+            var isStroke = !stroke.IsUndefined();
+            var streamGeometry = new StreamGeometry { FillRule = FillRule.Nonzero };
+            using (var sgc = streamGeometry.Open())
             {
-                gg.Children.Add(new EllipseGeometry(ToRect(rect)));
+                foreach (var rect in rectangles)
+                {
+                    const double arcAsBezier = 0.551784;
+
+                    var centerX = rect.Center.X;
+                    var centerY = rect.Center.Y;
+
+                    var midX = Math.Abs(rect.Width / 2.0) * arcAsBezier;
+                    var midY = Math.Abs(rect.Height / 2.0) * arcAsBezier;
+
+                    sgc.BeginFigure(new Point(rect.Right, centerY), isFilled, true);
+                    sgc.BezierTo(new Point(rect.Right, centerY + midY), new Point(centerX + midX, rect.Bottom), new Point(centerX, rect.Bottom), isStroke, true);
```

Comment:
I can see how this may be preferable for markers and such as a potential performance consideration, but I'd prefer there was an option like `UseFastEllipseApproximation` or something, and that this used arcs by default for precision. Is this measurably faster anyway? I'd guess that the savings from using streaming geometry are all that really matters here.

Reply:
Ah, you can't draw oblong ellipses with StreamingGeomtry can you, only Circles...

Reply:
I didn't really question this to be honest. You can actually draw arcs using `StreamGeometry`, so you should be able to draw a full ellipse exactly by drawing two arcs. I will give that a try and check if there is a noticeable performance difference.

Reply:
Ah, great: I didn't realise you could specify different radii for arcs.

    sgc.ArcTo(new Point(rect.Left, centerY), new Size(midX, midY), 180, false, SweepDirection.Clockwise, isStroke, true);
    sgc.ArcTo(new Point(rect.Right, centerY), new Size(midX, midY), 180, false, SweepDirection.Clockwise, isStroke, true);

I can't perceive much difference between the two; the Bezier's perhaps feel a little snappier; I've not tried to properly measure anything.

Reply:
Hi,
For the record, I didn't try using arcs and went straight for Bezier curves, because that's what is used internally in WPF to draw an ellipse geometry. Did you notice any noticeable inaccuracy using an ellipse geometry ?

Reply:
> that's what is used internally in WPF to draw an ellipse geometry

Could you provide a reference for that? I was looking for some implementation earlier, but couldn't find it (though I didn't look too hard). The `Arc` method seems to be produce the same results in the `CanvasRenderContext` as the 'old' `EllipseGeometry` based method does in the new `XamlRenderContext`.

> Did you notice any noticeable inaccuracy using an ellipse geometry

No, and for most ellipses it will be imperceptible, but for large ones it will mean differences between renderers and potentially hit-testing, and if precision is needed for some reason then it may not be acceptable. If it makes markers 2* faster, however, then there is an obvious benefit to be had ;)

Reply:
> 
> 
> > that's what is used internally in WPF to draw an ellipse geometry
> 
> Could you provide a reference for that? I was looking for some implementation earlier, but couldn't find it (though I didn't look too hard). The `Arc` method seems to be produce the same results in the `CanvasRenderContext` as the 'old' `EllipseGeometry` based method does in the new `XamlRenderContext`.

https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/EllipseGeometry.cs
All the path calculations start at line 272.
Although I chose a different factor ; but I have no strong feelings about it. IIRC the value used in WPF ensures all points are within the circle, the one I picked minimises the absolute error.

Reply:
Thanks for the link. It looks like [oblong arcs are approximated with Beziers anyway](Https://github.com/dotnet/wpf/blob/b697746306a695af80a756586cad3f84a62e366c/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/ArcSegment.cs#L76) - presumably as you describe - explaining why the `EllipseGeometry` and arc based streaming geometry look the same. This is somewhat upsetting, but oh well.

Reply:
Added an example for this in RenderingCapabilities. I don't think we need to worry too much about this.
![image](https://user-images.githubusercontent.com/18097859/95676692-2f4af500-0bc0-11eb-95f7-2afbe7f55387.png)

Reply:
For the record, I would prefer we use Arcs if there is no significant performance hit because it's simper, expresses the intention better, and is consistent with the `XAML` renderer; however, I agree that it won't be an issue for the typical use-cases.

Reply:
I guess you are right. It's also nice that the Arc-method doesn't need the 'magic' constant. I could not detect a significant performance difference in my (admittedly somewhat crude) benchmark.